// @aws-cdk/core/lib/cfn-parse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cfn_fn_1 = require("./cfn-fn");
const cfn_pseudo_1 = require("./cfn-pseudo");
const cfn_resource_policy_1 = require("./cfn-resource-policy");
const token_1 = require("./token");
/**
 * This class contains functions for translating from a pure CFN value
 * (like a JS object { "Ref": "Bucket" })
 * to a form CDK understands
 * (like Fn.ref('Bucket')).
 *
 * While this file not exported from the module
 * (to not make it part of the public API),
 * it is directly referenced in the generated L1 code,
 * so any renames of it need to be reflected in cfn2ts/codegen.ts as well.
 *
 * @experimental
 */
class FromCloudFormation {
    static parseValue(cfnValue) {
        return parseCfnValueToCdkValue(cfnValue);
    }
    // nothing to for any but return it
    static getAny(value) { return value; }
    // nothing to do - if 'value' is not a boolean or a Token,
    // a validator should report that at runtime
    static getBoolean(value) { return value; }
    static getDate(value) {
        // if the date is a deploy-time value, just return it
        if (token_1.isResolvableObject(value)) {
            return value;
        }
        // if the date has been given as a string, convert it
        if (typeof value === 'string') {
            return new Date(value);
        }
        // all other cases - just return the value,
        // if it's not a Date, a validator should catch it
        return value;
    }
    static getString(value) {
        // if the string is a deploy-time value, serialize it to a Token
        if (token_1.isResolvableObject(value)) {
            return value.toString();
        }
        // in all other cases, just return the input,
        // and let a validator handle it if it's not a string
        return value;
    }
    static getNumber(value) {
        // if the string is a deploy-time value, serialize it to a Token
        if (token_1.isResolvableObject(value)) {
            return token_1.Token.asNumber(value);
        }
        // in all other cases, just return the input,
        // and let a validator handle it if it's not a number
        return value;
    }
    static getStringArray(value) {
        // if the array is a deploy-time value, serialize it to a Token
        if (token_1.isResolvableObject(value)) {
            return token_1.Token.asList(value);
        }
        // in all other cases, delegate to the standard mapping logic
        return this.getArray(value, this.getString);
    }
    static getArray(value, mapper) {
        if (!Array.isArray(value)) {
            // break the type system, and just return the given value,
            // which hopefully will be reported as invalid by the validator
            // of the property we're transforming
            // (unless it's a deploy-time value,
            // which we can't map over at build time anyway)
            return value;
        }
        return value.map(mapper);
    }
    static getMap(value, mapper) {
        if (typeof value !== 'object') {
            // if the input is not a map (= object in JS land),
            // just return it, and let the validator of this property handle it
            // (unless it's a deploy-time value,
            // which we can't map over at build time anyway)
            return value;
        }
        const ret = {};
        for (const [key, val] of Object.entries(value)) {
            ret[key] = mapper(val);
        }
        return ret;
    }
    static parseDeletionPolicy(policy) {
        switch (policy) {
            case null: return undefined;
            case undefined: return undefined;
            case 'Delete': return cfn_resource_policy_1.CfnDeletionPolicy.DELETE;
            case 'Retain': return cfn_resource_policy_1.CfnDeletionPolicy.RETAIN;
            case 'Snapshot': return cfn_resource_policy_1.CfnDeletionPolicy.SNAPSHOT;
            default: throw new Error(`Unrecognized DeletionPolicy '${policy}'`);
        }
    }
    static getCfnTag(tag) {
        return tag == null
            ? {} // break the type system - this should be detected at runtime by a tag validator
            : {
                key: tag.Key,
                value: tag.Value,
            };
    }
}
exports.FromCloudFormation = FromCloudFormation;
function parseCfnValueToCdkValue(cfnValue) {
    // == null captures undefined as well
    if (cfnValue == null) {
        return undefined;
    }
    // if we have any late-bound values,
    // just return them
    if (token_1.isResolvableObject(cfnValue)) {
        return cfnValue;
    }
    if (Array.isArray(cfnValue)) {
        return cfnValue.map(el => parseCfnValueToCdkValue(el));
    }
    if (typeof cfnValue === 'object') {
        // an object can be either a CFN intrinsic, or an actual object
        const cfnIntrinsic = parseIfCfnIntrinsic(cfnValue);
        if (cfnIntrinsic) {
            return cfnIntrinsic;
        }
        const ret = {};
        for (const [key, val] of Object.entries(cfnValue)) {
            ret[key] = parseCfnValueToCdkValue(val);
        }
        return ret;
    }
    // in all other cases, just return the input
    return cfnValue;
}
function parseIfCfnIntrinsic(object) {
    var _a;
    const key = looksLikeCfnIntrinsic(object);
    switch (key) {
        case undefined:
            return undefined;
        case 'Ref': {
            // ToDo handle translating logical IDs
            return (_a = specialCaseRefs(object[key])) !== null && _a !== void 0 ? _a : cfn_fn_1.Fn._ref(object[key]);
        }
        case 'Fn::GetAtt': {
            // Fn::GetAtt takes a 2-element list as its argument
            const value = object[key];
            // ToDo same comment here as in Ref above
            return cfn_fn_1.Fn.getAtt((value[0]), value[1]);
        }
        case 'Fn::Join': {
            // Fn::Join takes a 2-element list as its argument,
            // where the first element is the delimiter,
            // and the second is the list of elements to join
            const value = parseCfnValueToCdkValue(object[key]);
            return cfn_fn_1.Fn.join(value[0], value[1]);
        }
        case 'Fn::If': {
            // Fn::If takes a 3-element list as its argument
            // ToDo the first argument is the name of the condition,
            // so we will need to retrieve the actual object from the template
            // when we handle preserveLogicalIds=false
            const value = parseCfnValueToCdkValue(object[key]);
            return cfn_fn_1.Fn.conditionIf(value[0], value[1], value[2]);
        }
        case 'Fn::Equals': {
            const value = parseCfnValueToCdkValue(object[key]);
            return cfn_fn_1.Fn.conditionEquals(value[0], value[1]);
        }
        default:
            throw new Error(`Unsupported CloudFormation function '${key}'`);
    }
}
function looksLikeCfnIntrinsic(object) {
    const objectKeys = Object.keys(object);
    // a CFN intrinsic is always an object with a single key
    if (objectKeys.length !== 1) {
        return undefined;
    }
    const key = objectKeys[0];
    return key === 'Ref' || key.startsWith('Fn::') ? key : undefined;
}
function specialCaseRefs(value) {
    switch (value) {
        case 'AWS::AccountId': return cfn_pseudo_1.Aws.ACCOUNT_ID;
        case 'AWS::Region': return cfn_pseudo_1.Aws.REGION;
        case 'AWS::Partition': return cfn_pseudo_1.Aws.PARTITION;
        case 'AWS::URLSuffix': return cfn_pseudo_1.Aws.URL_SUFFIX;
        case 'AWS::NotificationARNs': return cfn_pseudo_1.Aws.NOTIFICATION_ARNS;
        case 'AWS::StackId': return cfn_pseudo_1.Aws.STACK_ID;
        case 'AWS::StackName': return cfn_pseudo_1.Aws.STACK_NAME;
        case 'AWS::NoValue': return cfn_pseudo_1.Aws.NO_VALUE;
        default: return undefined;
    }
}
