// @aws-cdk/aws-ec2/test/test.machine-image.js
"use strict";
const core_1 = require("@aws-cdk/core");
const ec2 = require("../lib");
let app;
let stack;
function isWindowsUserData(ud) {
    return ud.render().indexOf('powershell') > -1;
}
function isLinuxUserData(ud) {
    return ud.render().indexOf('bash') > -1;
}
module.exports = {
    'setUp'(cb) {
        app = new core_1.App();
        stack = new core_1.Stack(app, 'Stack', {
            env: { account: '1234', region: 'testregion' },
        });
        cb();
    },
    'can make and use a Windows image'(test) {
        // WHEN
        const image = new ec2.GenericWindowsImage({
            testregion: 'ami-1234',
        });
        // THEN
        const details = image.getImage(stack);
        test.equals(details.imageId, 'ami-1234');
        test.equals(details.osType, ec2.OperatingSystemType.WINDOWS);
        test.done();
    },
    'WindowsImage retains userdata if given'(test) {
        // WHEN
        const ud = ec2.UserData.forWindows();
        const image = new ec2.GenericWindowsImage({
            testregion: 'ami-1234',
        }, {
            userData: ud,
        });
        // THEN
        const details = image.getImage(stack);
        test.equals(details.userData, ud);
        test.done();
    },
    'WindowsImage creates UserData if not given'(test) {
        // WHEN
        const image = new ec2.GenericWindowsImage({
            testregion: 'ami-1234',
        });
        // THEN
        const details = image.getImage(stack);
        test.ok(isWindowsUserData(details.userData));
        test.done();
    },
    'LookupMachineImage default search'(test) {
        // GIVEN
        // WHEN
        new ec2.LookupMachineImage({ name: 'bla*', owners: ['amazon'] }).getImage(stack);
        // THEN
        const missing = app.synth().manifest.missing || [];
        test.deepEqual(missing, [
            {
                key: 'ami:account=1234:filters.image-type.0=machine:filters.name.0=bla*:filters.state.0=available:owners.0=amazon:region=testregion',
                props: {
                    account: '1234',
                    region: 'testregion',
                    owners: ['amazon'],
                    filters: {
                        'name': ['bla*'],
                        'state': ['available'],
                        'image-type': ['machine'],
                    },
                },
                provider: 'ami',
            },
        ]);
        test.done();
    },
    'LookupMachineImage creates correct type of UserData'(test) {
        // WHEN
        const linuxDetails = new ec2.LookupMachineImage({ name: 'bla*', owners: ['amazon'] }).getImage(stack);
        const windowsDetails = new ec2.LookupMachineImage({ name: 'bla*', owners: ['amazon'], windows: true }).getImage(stack);
        // THEN
        test.ok(isWindowsUserData(windowsDetails.userData));
        test.ok(isLinuxUserData(linuxDetails.userData));
        test.done();
    },
};
