// @aws-cdk/core/lib/stack-synthesizers/nested.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Deployment environment for a nested stack
 *
 * Interoperates with the StackSynthesizer of the parent stack.
 */
class NestedStackSynthesizer {
    constructor(parentDeployment) {
        this.parentDeployment = parentDeployment;
    }
    bind(_stack) {
        // Nothing to do
    }
    addFileAsset(asset) {
        // Forward to parent deployment. By the magic of cross-stack references any parameter
        // returned and used will magically be forwarded to the nested stack.
        return this.parentDeployment.addFileAsset(asset);
    }
    addDockerImageAsset(asset) {
        // Forward to parent deployment. By the magic of cross-stack references any parameter
        // returned and used will magically be forwarded to the nested stack.
        return this.parentDeployment.addDockerImageAsset(asset);
    }
    synthesizeStackArtifacts(_session) {
        // Do not emit Nested Stack as a cloud assembly artifact.
        // It will be registered as an S3 asset of its parent instead.
    }
}
exports.NestedStackSynthesizer = NestedStackSynthesizer;
