// @aws-cdk/cdk-assets-schema/lib/private/schema-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const file_asset_1 = require("../file-asset");
/**
 * Validate that a given key is of a given type in an object
 *
 * If not optional, the key is considered required.
 *
 * Uses predicate validators that return a 'witness', so we can use the return type of
 * the validator function to infer the actual type of the value and enrich the type information
 * of the given input field.
 *
 * In effect, validators should be written like this:
 *
 *   if (!valid(input)) { throw; }
 *   return input;
 */
function expectKey(obj, key, validate, optional) {
    if (typeof obj !== 'object' || obj === null || (!(key in obj) && !optional)) {
        throw new Error(`Expected key '${key}' missing: ${JSON.stringify(obj)}`);
    }
    if (key in obj) {
        try {
            validate(obj[key]);
        }
        catch (e) {
            throw new Error(`${key}: ${e.message}`);
        }
    }
}
exports.expectKey = expectKey;
function isString(x) {
    if (typeof x !== 'string') {
        throw new Error(`Expected a string, got '${x}'`);
    }
    return x;
}
exports.isString = isString;
function isMapOf(pred) {
    return x => {
        assertIsObject(x);
        Object.values(x).forEach(pred);
        return x;
    };
}
exports.isMapOf = isMapOf;
function isObjectAnd(p) {
    return x => {
        assertIsObject(x);
        return p(x);
    };
}
exports.isObjectAnd = isObjectAnd;
function assertIsObject(x) {
    if (typeof x !== 'object' || x === null) {
        throw new Error(`Expected a map, got '${x}'`);
    }
}
exports.assertIsObject = assertIsObject;
function isFileAssetPackaging(x) {
    const str = isString(x);
    const validValues = Object.values(file_asset_1.FileAssetPackaging); // Explicit cast needed because this is a string-valued enum
    if (!validValues.includes(str)) {
        throw new Error(`Expected a FileAssetPackaging (one of ${validValues.map(v => `'${v}'`).join(', ')}), got '${str}'`);
    }
    return x;
}
exports.isFileAssetPackaging = isFileAssetPackaging;
